File: C/stack/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "stack.h"
#include "../array_util.h"
#include "../error.h"

Stack* new_stack(void)
{
	return new_stack_with_capacity(DEFAULT_CAPACITY);
}

Stack* new_stack_with_capacity(int initial_capacity)
{
	if (initial_capacity <= 0)
	{
		exit_with_error("new_stack - Capacity must be greater than 0");
	}

	Stack* stack = malloc(sizeof(Stack));
	if (stack == NULL)
	{
		exit_with_error("new_stack - Out of memory");
	}

	stack->size = 0;
	stack->capacity = initial_capacity;
	stack->array = create_array(stack->capacity);

	return stack;
}

void delete_stack(Stack* stack)
{
	if (stack != NULL)
	{
		free(stack->array);
		free(stack);
	}
}

Stack* copy_stack(const Stack* other)
{
	if (other == NULL)
	{
		return NULL;
	}

	Stack* stack = new_stack_with_capacity(other->capacity);

	stack->size = other->size;
	memcpy(stack->array, other->array, other->size * sizeof(int));

	return stack;
}


bool push(Stack* stack, int value)
{
	if (stack == NULL)
	{
		exit_with_error("push - stack is null");
	}

	//double the capacity if full
	if (isFull(stack))
	{
		expand_capacity(stack, stack->capacity * 2);
	}

	stack->array[stack->size] = value;
	stack->size++;

	return true;
}

int pop(Stack* stack)
{
	if (stack == NULL)
	{
		exit_with_error("pop - stack is null");
	}
	else if (isEmpty(stack))
	{
		exit_with_error("pop - stack is empty");
	}

	stack->size--;
	return stack->array[stack->size];
}

int peek(const Stack* stack)
{
	if (stack == NULL)
	{
		exit_with_error("pop - stack is null");
	}
	else if (isEmpty(stack))
	{
		exit_with_error("peek - stack is empty");
	}

	return stack->array[stack->size - 1];
}

bool isEmpty(const Stack* stack)
{
	return stack == NULL || stack->size == 0;
}

bool isFull(const Stack* stack)
{
	return stack != NULL && stack->size == stack->capacity;
}

void clear(Stack* stack)
{
	if (stack == NULL)
	{
		exit_with_error("clear - stack is null");
	}

	stack->size = 0;
}

bool expand_capacity(Stack* stack, int new_capacity)
{
	//TODO TRUNCATE IF NEW CAPACITY IS SMALLER?
	//TODO WHAT IF SIZE IS TOO LARGE?

	if (stack == NULL)
	{
		exit_with_error("expand_capacity - stack is null");
	}

	if (new_capacity <= stack->capacity)
	{
		return false;
	}

	int* new_array = copy_array(stack->array, stack->size, new_capacity);

	//replace the old array
	free(stack->array);
	stack->array = new_array;
	stack->capacity = new_capacity;

	return true;
}

void dump(const Stack* stack)
{
	if (stack == NULL)
	{
		printf("Stack is null.\n");
		return;
	}

	print_array(stack->array, stack->size, stack->capacity);
}

File: Java/stack/Stack.java
import java.util.*;

/**
 * Array implementation of a stack
 */
public class Stack<T> extends ArrayCollection<T>
{
	/**
	 * Creates a new stack
	 */
	public Stack()
	{
		super();
	}

	/**
	 * Creates a new stack with a specified initial capacity
	 */
	public Stack(int initialCapacity)
	{
		super(initialCapacity);
	}

	/**
	 * Copies the stack
	 * 
	 * @param other the stack to copy
	 */
	public Stack(Stack<T> other)
	{
		super(other);
	}

	/**
	 * Adds an item to the stack.
	 * Expands capacity if full.
	 * @param item the item to add to the stack
	 */
	public void push(T item)
	{
		if (isFull())
		{
			expandCapacity(array.length * 2);
		}

		array[size] = item;
		size++;
	}

	/**
	 * Returns and removes the item on top of the stack
	 * 
	 * @return the item on top of the stack
	 * @throws EmptyStackException if the stack is empty
	 */
	public T pop()
	{
		if (isEmpty())
		{
			throw new EmptyStackException();
		}

		T item = array[size-1];
		size--;
		return item;
	}

	/**
	 * @return the item on top of the stack without removing it
	 * @throws EmptyStackException if the stack is empty
	 */
	public T peek()
	{
		if (isEmpty())
		{
			throw new EmptyStackException();
		}

		return array[size-1];
	}

	@Override
	public boolean equals(Object other)
	{
		if (this == other)
		{
			return true;
		}

		if (other == null)
		{
			return false;
		}

		if (!(other instanceof Stack<?>))
		{
			return false;
		}

		if (size != ((Stack<?>)other).getSize())
		{
			return false;
		}

		for (int i = 0; i < size; i++)
		{
			if (!array[i].equals(((Stack<?>)other).array[i]))
			{
				return false;
			}
		}

		return true;
	}
}
